### ISSUE TITLE
Filtering a collection on a custom attribute and a custom store_id generate wrong sql request

### ISSUE THREAD

**Issue by alexmtch** [magento/magento2#23720]
### Preconditions (*)
1. Magento CE 2.3.2
2. Php 7.1
3. developer (or production ?) mode

### Steps to reproduce (*)
1. Create a custom attribute with a store_id scope
2. Create a product, attach it to a new store view (not the website or the store group) and set a value for the specific store to the custom attribute
3. Create a product collection on a custom attribute following this code :
```
$products = $this->productFactory->create()
                ->getCollection()
                ->addStoreFilter(2)
                ->addAttributeToFilter('custom_attribute', '123');
```

### Expected result (*)
1. In my case i expected a result of a collection which contains 1 item

### Actual result (*)
1. I got an empty collection

After a workaround i retrieve the sql request
```
SELECT `e`.*, IF(at_custom_attribute.value_id > 0, at_custom_attribute.value, at_custom_attribute_default.value) AS `custom_attribute` 
FROM `catalog_product_entity` AS `e`
INNER JOIN `catalog_product_website` AS `product_website` ON product_website.product_id = e.entity_id AND product_website.website_id = 2
INNER JOIN `catalog_product_entity_varchar` AS `at_custom_attribute_default` ON (`at_custom_attribute_default`.`entity_id` = `e`.`entity_id`) AND (`at_custom_attribute_default`.`attribute_id` = '163') AND `at_custom_attribute_default`.`store_id` = 0
LEFT JOIN `catalog_product_entity_varchar` AS `at_custom_attribute` ON (`at_custom_attribute`.`entity_id` = `e`.`entity_id`) AND (`at_custom_attribute`.`attribute_id` = '163') AND (`at_custom_attribute`.`store_id` = 2) WHERE (IF(at_custom_attribute.value_id > 0, at_custom_attribute.value, at_custom_attribute_default.value) = '123')
```

You see on the line 4 a join condition on the store id where the value is 0 while it should be 2. I fixed the sql request and run into mysql and i got the correct sql result.

After a workaround this is generated by the method _joinAttributeToSelect of this file vendor/magento/module-catalog/Model/ResourceModel/Collection/AbstractCollection.php

### Update
I see in this method this comment

> /**
> * Add joining default value for not default store
> * if value for store is null - we use default value
> */

I suppose by this comment that this declaration
```
$defCondition .= $connection->quoteInto(
                " AND " . $connection->quoteColumnAs("{$defAlias}.store_id", null) . " = ?",
                $this->getDefaultStoreId()
            );
```
should be
```
$defCondition .= $connection->quoteInto(
                " AND " . $connection->quoteColumnAs("{$defAlias}.store_id", null) . " = ?",
                $storeId ? $storeId : $this->getDefaultStoreId()
            );
```
What do you think about this ?

**Comment by magento-engcom-team** [comment]
@engcom-Bravo Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [x] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-18204` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by alexmtch** [comment]
I am not enough experimented with Github to know how to make a pull request based on this issue so i let other contributors/maintainers create the fix based on the observations i put in the post.

**Comment by Nazar65** [comment]
@alexmtch i don't think that this an issue because you set store_id not store_view_id

**Comment by sunilit42** [comment]
@alexmtch  

That is right query, it will check first store wise if not found then it will check globally

**Comment by alexmtch** [comment]
@sunilit42 

The first condition 242-246 is right but the second at 250 won't works with store view.

You have 2 stores view
- store A (id = 2)
- store B (id = 3)

You have an attribute named custom_attribute and its scope is store view

When you process this
```
$products = $this->productFactory->create()
                ->getCollection()
                ->addStoreFilter(2)
                ->addAttributeToFilter('custom_attribute', '123');
```

After the 242-246 check $storeId value is 2 but whith the condition at 250
`if ($storeId != $this->getDefaultStoreId() && !$attribute->isScopeGlobal())`

The $storeId value is different from 0 and the attribute isn't sope global so it run in and force the storeId to 0 which is an error.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-23837` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by alexmtch** [comment]
Hello @engcom-Hotel,

Why doest it takes almost 5 years to change `$this->getDefaultStoreId()` to `$storeId ? $storeId : $this->getDefaultStoreId()` ?

**Comment by engcom-Hotel** [comment]
Hello @alexmtch,

Thank you for reaching out to us!

We noticed that the internal JIRA [AC-764](https://jira.corp.adobe.com/browse/AC-764) ticket has not yet been picked up by any team members, possibly because the team is currently occupied with high-priority tickets.

We always welcome contributions from the community. Please feel free to raise a new pull request (PR) with the necessary fixes, and we will be happy to review and process it further.

Thanks
